### ISSUE TITLE
🎃 Convert Service Classes to Static Fields 🎃

### ISSUE THREAD

**Issue by calebcartwright** [badges/shields#5555]
We recently completed the [upgrade to Node 12 in our production environment](https://github.com/badges/shields/pull/5436) which, among other benefits, now allows us to leverage [static fields](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Public_class_fields) for defining core attributes like the category and route. 

We've completed the refactoring of several service classes to use static fields, but there's a lot of them remaining and we could use your help to convert the rest!

## What are the steps required?

This work involves updating the members in a service class (found in the *-service.js files) to convert the static getters to static fields,

For example:
```js
  static get category() {
    return 'version'
  }
```
 
becomes
```js
  static category = 'version'
```

If a service class doesn't have any getters (e.g. redirector classes like [this one](https://github.com/badges/shields/blob/master/services/offset-earth/offset-earth-trees-redirect.service.js), then there's nothing needed for that class!

Additional information and references can be found in the corresponding section of the [tutorial documentation](https://github.com/badges/shields/blob/master/doc/TUTORIAL.md#42-our-first-badge) as well previous PRs that did this conversion like #5552 or #5530 

## How to contribute?

Below is the list of remaining service class families (as a link to their location in source) that need to be refactored. If you're interested in working on this, leave a comment on this thread or drop us a note in [Discord](https://discord.gg/HjJCwm5) letting us know which one(s) you'd like to work on. Items that have been checked off have already been completed, but the others are up for grabs!

As a rule of thumb, we'd generally prefer to see PRs contain between 2-4 refactored service class families to keep the review and merge process streamlined. You can work on as many as you'd like, but please try to follow that guidance when contributing and split PRs accordingly!

- [x] [gemnasium](https://github.com/badges/shields/tree/master/services/gemnasium)
- [x] [gitlab](https://github.com/badges/shields/tree/master/services/gitlab)
- [x] [gitter](https://github.com/badges/shields/tree/master/services/gitter)
- [x] [gratipay](https://github.com/badges/shields/tree/master/services/gratipay)
- [x] [hackage](https://github.com/badges/shields/tree/master/services/hackage)
- [x] [hexpm](https://github.com/badges/shields/tree/master/services/hexpm)
- [x] [homebrew](https://github.com/badges/shields/tree/master/services/homebrew)
- [x] [hsts](https://github.com/badges/shields/tree/master/services/hsts)
- [x] [imagelayers](https://github.com/badges/shields/tree/master/services/imagelayers)
- [x] [issuestats](https://github.com/badges/shields/tree/master/services/issuestats)
- [x] [itunes](https://github.com/badges/shields/tree/master/services/itunes)
- [x] [jenkins](https://github.com/badges/shields/tree/master/services/jenkins)
- [x] [jetbrains](https://github.com/badges/shields/tree/master/services/jetbrains)
- [x] [jira](https://github.com/badges/shields/tree/master/services/jira)
- [x] [jitpack](https://github.com/badges/shields/tree/master/services/jitpack)
- [x] [jsdelivr](https://github.com/badges/shields/tree/master/services/jsdelivr)
- [x] [keybase](https://github.com/badges/shields/tree/master/services/keybase)
- [x] [leanpub](https://github.com/badges/shields/tree/master/services/leanpub)
- [x] [lgtm](https://github.com/badges/shields/tree/master/services/lgtm) 
- [x] [liberapay](https://github.com/badges/shields/tree/master/services/liberapay)
- [x] [librariesio](https://github.com/badges/shields/tree/master/services/librariesio)
- [x] [libscore](https://github.com/badges/shields/tree/master/services/libscore)
- [x] [localizely](https://github.com/badges/shields/tree/master/services/localizely)
- [x] [luarocks](https://github.com/badges/shields/tree/master/services/luarocks)
- [x] [magnumci](https://github.com/badges/shields/tree/master/services/magnumci)
- [x] [maintenance](https://github.com/badges/shields/tree/master/services/maintenance)
- [x] [mastodon](https://github.com/badges/shields/tree/master/services/mastodon)
- [x] [matrix](https://github.com/badges/shields/tree/master/services/matrix)
- [x] [maven-central](https://github.com/badges/shields/tree/master/services/maven-central)
- [x] [maven-metadata](https://github.com/badges/shields/tree/master/services/maven-metadata)
- [x] [microbadger](https://github.com/badges/shields/tree/master/services/microbadger)
- [x] [mozilla-observatory](https://github.com/badges/shields/tree/master/services/mozilla-observatory)
- [x] [myget](https://github.com/badges/shields/tree/master/services/myget)
- [x] [netlify](https://github.com/badges/shields/tree/master/services/netlify) 
- [x] [nexus](https://github.com/badges/shields/tree/master/services/nexus)
- [x] [node](https://github.com/badges/shields/tree/master/services/node)
- [x] [nodeping](https://github.com/badges/shields/tree/master/services/nodeping)
- [x] [npm](https://github.com/badges/shields/tree/master/services/npm)
- [x] [nsp](https://github.com/badges/shields/tree/master/services/nsp)
- [x] [nuget](https://github.com/badges/shields/tree/master/services/nuget)
- [x] [nycrc](https://github.com/badges/shields/tree/master/services/nycrc)
- [x] [offset-earth](https://github.com/badges/shields/tree/master/services/offset-earth)
- [x] [opencollective](https://github.com/badges/shields/tree/master/services/opencollective)
- [x] [opm](https://github.com/badges/shields/tree/master/services/opm)
- [x] [osslifecycle](https://github.com/badges/shields/tree/master/services/osslifecycle)
- [x] [packagecontrol](https://github.com/badges/shields/tree/master/services/packagecontrol)
- [x] [packagist](https://github.com/badges/shields/tree/master/services/packagist)
- [x] [php-eye](https://github.com/badges/shields/tree/master/services/php-eye)
- [x] [pkgreview](https://github.com/badges/shields/tree/master/services/pkgreview)
- [x] [poeditor](https://github.com/badges/shields/tree/master/services/poeditor)
- [x] [powershellgallery](https://github.com/badges/shields/tree/master/services/powershellgallery)
- [x] [pub](https://github.com/badges/shields/tree/master/services/pub)
- [x] [puppetforge](https://github.com/badges/shields/tree/master/services/puppetforge)
- [x] [pypi](https://github.com/badges/shields/tree/master/services/pypi)
- [x] [readthedocs](https://github.com/badges/shields/tree/master/services/readthedocs)
- [x] [reddit](https://github.com/badges/shields/tree/master/services/reddit)
- [x] [redmine](https://github.com/badges/shields/tree/master/services/redmine)
- [x] [repology](https://github.com/badges/shields/tree/master/services/repology)
- [x] [requires](https://github.com/badges/shields/tree/master/services/requires)
- [x] [resharper](https://github.com/badges/shields/tree/master/services/resharper)
- [x] [scoop](https://github.com/badges/shields/tree/master/services/scoop)
- [x] [scrutinizer](https://github.com/badges/shields/tree/master/services/scrutinizer)
- [x] [security-headers](https://github.com/badges/shields/tree/master/services/security-headers)
- [x] [shippable](https://github.com/badges/shields/tree/master/services/shippable)
- [x] [snap-ci](https://github.com/badges/shields/tree/master/services/snap-ci)
- [x] [snyk](https://github.com/badges/shields/tree/master/services/snyk)
- [x] [sonar](https://github.com/badges/shields/tree/master/services/sonar)
- [x] [sourceforge](https://github.com/badges/shields/tree/master/services/sourceforge)
- [x] [sourcegraph](https://github.com/badges/shields/tree/master/services/sourcegraph)
- [x] [spack](https://github.com/badges/shields/tree/master/services/spack)
- [x] [spiget](https://github.com/badges/shields/tree/master/services/spiget)
- [x] [stackexchange](https://github.com/badges/shields/tree/master/services/stackexchange)
- [x] [static-badge](https://github.com/badges/shields/tree/master/services/static-badge)
- [x] [steam](https://github.com/badges/shields/tree/master/services/steam)
- [x] [swagger](https://github.com/badges/shields/tree/master/services/swagger)
- [x] [symfony](https://github.com/badges/shields/tree/master/services/symfony)
- [x] [teamcity](https://github.com/badges/shields/tree/master/services/teamcity)
- [x] [travis](https://github.com/badges/shields/tree/master/services/travis)
- [x] [treeware](https://github.com/badges/shields/tree/master/services/treeware)
- [x] [twitch](https://github.com/badges/shields/tree/master/services/twitch)
- [x] [twitter](https://github.com/badges/shields/tree/master/services/twitter)
- [x] [ubuntu](https://github.com/badges/shields/tree/master/services/ubuntu)
- [x] [uptimerobot](https://github.com/badges/shields/tree/master/services/uptimerobot)
- [x] [vaadin-directory](https://github.com/badges/shields/tree/master/services/vaadin-directory)
- [x] [versioneye](https://github.com/badges/shields/tree/master/services/versioneye)
- [x] [visual-studio-app-center](https://github.com/badges/shields/tree/master/services/visual-studio-app-center)
- [x] [visual-studio-marketplace](https://github.com/badges/shields/tree/master/services/visual-studio-marketplace)
- [x] [w3c](https://github.com/badges/shields/tree/master/services/w3c)
- [x] [waffle](https://github.com/badges/shields/tree/master/services/waffle)
- [ ] [website](https://github.com/badges/shields/tree/master/services/website) (WIP @monizb)
- [ ] [wercker](https://github.com/badges/shields/tree/master/services/wercker) (WIP @monizb)
- [ ] [wheelmap](https://github.com/badges/shields/tree/master/services/wheelmap) (WIP @monizb)
- [ ] [wordpress](https://github.com/badges/shields/tree/master/services/wordpress) (WIP @monizb)
- [ ] [youtube](https://github.com/badges/shields/tree/master/services/youtube) (WIP @monizb)

**Comment by jabbar86** [comment]
@calebcartwright I would like to contribute to this. Should I wait for hacktoberfest to start?

**Comment by ThakurKarthik** [comment]
Hi @calebcartwright I would love to work on some classes.

**Comment by calebcartwright** [comment]
@jabbar86 

>  I would like to contribute to this.

Great! Please specify which of the service classes families you're going to work on so that we can prevent any duplicative efforts

> Should I wait for hacktoberfest to start?

that's up to you! We'll happily accept help with this effort regardless of what month it is, but PRs submitted before Hacktoberfest begins will obviously not count towards your Hacktoberfest contributions

@ThakurKarthik 

> I would love to work on some classes.I have posted the same message in discord with username cosmos

Great! I've updated the above list to reflect the ones you've claimed

**Comment by jabbar86** [comment]
@calebcartwright Thanks I will contribute before hacktoberfest too. I will be working on.

```
homebrew
hsts
itunes
jenkins
jetbrains
```

**Comment by ThakurKarthik** [comment]
@calebcartwright can i work on this  jira jitpack jsdelivr keybase , if no one else is working on those?

**Comment by jabbar86** [comment]
@calebcartwright I have finish https://github.com/badges/shields/issues/5555#issuecomment-695256771. I would like to take another modules i.e 

```
lgtm
liberapay
librariesio
localizely
luarocks
```

**Comment by jabbar86** [comment]
@calebcartwright I have completed https://github.com/badges/shields/issues/5555#issuecomment-695775615. Assign me 

```
maven-metadata
microbadger
mozilla-observatory
myget
netlify
```

If anyone hasn't started working on it.Thanks

**Comment by jabbar86** [comment]
@calebcartwright Assign Me 
```

nexus
node
nodeping
npm
nsp
```

**Comment by ThakurKarthik** [comment]
I am working on these services [nuget nycrc opencollective opm osslifecycle]

**Comment by monizb** [comment]
@calebcartwright I would like to take up:

```
website
wercker
youtube
wheelmap
wordpress
```
Thanks :)

**Comment by vanhaaggen** [comment]
@calebcartwright can i help out with these? 

`snyk | sonar | sourceforge | sourcegraph | spack`

**Comment by ThakurKarthik** [comment]
Hey @calebcartwright if no one taken ```[ osslifecycle packagecontrol packagist pkgreview poeditor ]``` already I would like to work on them. Thanks !

**Comment by RDIL** [comment]
@calebcartwright I can take any of the ones starting with `s`, `t`, `u`, or `v`.

**Comment by ThakurKarthik** [comment]
@RDIL feel free to pick any 4 - 5 services which are not done and post them here or at shields discord channel  to prevent duplicate work.Thanks !

**Comment by ThakurKarthik** [comment]
@calebcartwright I am taking these files next  [ powershellgallery pub puppetforge pypi ].

**Comment by ThakurKarthik** [comment]
@calebcartwright  I have changes for the above services also.Meanwhile I am working on these services [readthedocs reddit redmine repology] and open PR once the previous changes get's accepted. Thanks !

**Comment by ThakurKarthik** [comment]
I am working on these too [ requires resharper scoop scrutinizer] Thanks !

**Comment by jabbar86** [comment]
@calebcartwright I will be working in 2 parts so that its easy to maintain the code.

1st part 
```
swagger
symfony
teamcity
travis
treeware
```

2nd part 
```

twitch
twitter
ubuntu
uptimerobot
vaadin-directory
```
Thanks.

**Comment by BalgopalPatro** [comment]
@calebcartwright 
I wanna do for the followings
visual-studio-app-center
visual-studio-marketplace
w3c

**Comment by MonliH** [comment]
I would like to help as well, although it seems all the badges have been "claimed".

**Comment by calebcartwright** [comment]
@MonliH - so it seems! If you're interested there's **tons** of other Shields-related work in need of assistance

https://github.com/badges/shields/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22

**Comment by ThakurKarthik** [comment]
Hi @calebcartwright quick update here this PR https://github.com/badges/shields/pull/5635 already fixed the [powershellgallery, 
 pub, puppetforge, pypi] services. You can mark them done. Thanks !

**Comment by BalgopalPatro** [comment]
> @calebcartwright
> I wanna do for the followings
> visual-studio-app-center
> visual-studio-marketplace
> w3c

@calebcartwright Please review completed task with PR #5642

**Comment by PyvesB** [comment]
Things are getting along quite nicely, thanks to all those involved so far! 🙌🏻

**Comment by JoeIzzard** [comment]
@monizb I didn't see this thread until after I had submitted the PR's. Wordpress has been done, partly as I was adding badges and partly as it's own PR. Related PR's:
- #5493 - This was split into multiple PR's (below), originally tackled some (Not sure if all) here
- #5717 - Part refactored whilst adding badges
- #5725 - Remaining parts of WordPress service

Sorry for the inconvenience! Hopefully it's not wasted any work you've done.

**Comment by jabbar86** [comment]
@JoeIzzard if it's still not completed I can take up this task.

**Comment by JoeIzzard** [comment]
@jabbar86 Those PR's should resolve the WordPress service completely.

**Comment by geekyrajshri** [comment]
@calebcartwright  I would like to contribute on any tasks left on this. Let me know if that's available

**Comment by calebcartwright** [comment]
Thanks @geekyrajshri! At this point all the service classes enumerated above have either been resolved or are already pending by another user. However, I do believe there's a handful of service classes in the GitHub family that I thought I would have already finished when I made the original list in the issue description, but I didn't end up finishing theme all.

That means that most/all of the service classes inclusively from GitHubSearch (https://github.com/badges/shields/blob/master/services/github/github-search.service.js) through GitHubWorkFlowStatus probably still need to be refactored if you (or anyone else) wants to work on those

**Comment by vividh** [comment]
Wondering why a codemod wasn't used for this?

**Comment by calebcartwright** [comment]
> Wondering why a codemod wasn't used for this?

@vividh - what exactly is the motivation for the question, are you just curious about why this work was completed the way it was? The short answer is that it made more sense for us to solicit assistance from our community of contributors vs a single maintainer figuring out how to automate everything. 

We review each and every change in detail to maintain the stability, reliability, and quality of the Shields.io service. And while the nature of these changes is rather trivial, we also knew it was never going to be feasible to update and review all of these files in one pass because those trivial changes still produce large diffs.

As such we also knew this work was going to have to be done over many PRs and not something we wanted to stick with a single person, hence the call for assistance from the community.

From the perspective of the Shields maintainers, it doesn't really matter to us _how_ folks helping us with this issue made the changes (whether manually by hand or automation), but it definitely wasn't going to be an effective use of our time to research, select, document, and recommend a codemod/AST-based automation approach for individual contributors to attempt to adopt.. Especially since most of these contributors would likely have found that to be more time consuming and challenging than tweaking 4 or 5 files by hand anyway.

**Comment by vividh** [comment]
Thanks for the clarification @calebcartwright. My question was borne out of curiosity, since I didn't see any discussions of codemods on this issue, and wanted to understand if it was considered.

The way you put it though makes sense to me. Codemods don't have the best documentation, and recommending to use that would possibly have alienated many potential contributors.

**Comment by calebcartwright** [comment]
@monizb - are you still working on the last set of service class files as you [previously indicated](https://github.com/badges/shields/issues/5555#issuecomment-700420649)? We'd love to get this wrapped up before October ends if possible!

**Comment by vividh** [comment]
I can take these up if @monizb is unable to.

**Comment by homersimpsons** [comment]
@calebcartwright This issue might be done as #5793 is merged.

I did mention in the Pull Request
> Searching for `static get ` still gives some results that could be migrated to `static variable = ...`

Should I address the remaining `static`s ?

**Comment by calebcartwright** [comment]
> @calebcartwright This issue might be done as #5793 is merged.

Thanks for the follow up @homersimpsons! 

> Should I address the remaining statics ?

We wanted to address this refactoring in buckets, with the first being `core` (which is done) and then the service classes. There's  a few static getters on the `master` branch in the package (under `badge-maker`) but do **not** bother with any of those, as there's a separate sizeable refactoring of that code inflight that has already done so (refs #5754).

There does appear to be a handful of service classes that were missed altogether for one reason or another, like `aur` and `jenkins-coverage`, as well as a few fields that were missed in classes that were refactored. If you'd like to work on any of those returning a static value then that'd be great! Be sure to split changes up into multiple PRs once the diffs start to get over a few hundred lines to make them easier to review.

Please also note though that Discord was intentionally skipped for now to minimize merge conflicts with #5079, and that not every single `static get ` instance can/should be converted, for example:

https://github.com/badges/shields/blob/5e335e1bb2601f3cc27aee9c67e5a7fe9e309a3c/services/youtube/youtube-comments.service.js#L11-L25

**Comment by paulmelnikow** [comment]
> There's a few static getters on the `master` branch in the package (under `badge-maker`) but do **not** bother with any of those

Just a heads up that they also have to be left for now because we still support Node 10 for the package.

**Comment by calebcartwright** [comment]
Thanks to everyone that helped with this work! Seems all the service class files have been refactored now, so going to go ahead and close.
